src/Compilers/CSharp/Test/Semantic/SourceGeneration/SyntaxAwareGeneratorTests.cs

```diff
@@ -83,22 +84,130 @@ class C { }
 
             Assert.Single(compilation.SyntaxTrees);
 
+            ISyntaxContextReceiver? receiver = null;
+
             var testGenerator = new CallbackGenerator(
-                onInit: initialize,
-                onExecute: (e) => { }
+                onInit: (i) => i.RegisterForSyntaxNotifications(() => new TestSyntaxContextReceiver()),
+                onExecute: (e) => receiver = e.SyntaxContextReceiver
                 );
 
             GeneratorDriver driver = CSharpGeneratorDriver.Create(new[] { testGenerator }, parseOptions: parseOptions);
             driver.RunGeneratorsAndUpdateCompilation(compilation, out _, out _);
 
-            void initialize(GeneratorInitializationContext initContext)
+            Assert.NotNull(receiver);
+            Assert.IsType<TestSyntaxContextReceiver>(receiver);
+        }
+
+        [Fact]
+        public void SyntaxContext_Receiver_Is_Null_WhenNot_Registered()
+        {
+            var source = @"
+class C { }
+";
+            var parseOptions = TestOptions.Regular;
+            Compilation compilation = CreateCompilation(source, options: TestOptions.DebugDll, parseOptions: parseOptions);
+            compilation.VerifyDiagnostics();
+
+            Assert.Single(compilation.SyntaxTrees);
+
+            ISyntaxContextReceiver? receiver = null;
+
+            var testGenerator = new CallbackGenerator(
+                onInit: (i) => { },
+                onExecute: (e) => receiver = e.SyntaxContextReceiver
+                );
+
+            GeneratorDriver driver = CSharpGeneratorDriver.Create(new[] { testGenerator }, parseOptions: parseOptions);
+            driver.RunGeneratorsAndUpdateCompilation(compilation, out var outputCompilation, out _);
+
+            Assert.Null(receiver);
+        }
+
+        [Fact]
+        public void SyntaxContext_Receiver_Is_Null_When_Syntax_Receiver_Registered()
+        {
+            var source = @"
+class C { }
+";
+            var parseOptions = TestOptions.Regular;
+            Compilation compilation = CreateCompilation(source, options: TestOptions.DebugDll, parseOptions: parseOptions);
+            compilation.VerifyDiagnostics();
+
+            Assert.Single(compilation.SyntaxTrees);
+
+            ISyntaxContextReceiver? receiver = null;
```

Comment:
Consider having two locals here: `ISyntaxContextReceiver` and `ISyntaxReceiver` then read them both in the lambda callback. That way the `Assert` at the end can be a pair of `Null` and `NotNull`. The presence of `NotNull` ensures the lambda is executing as expected where as the current form of the test could succeed if the lambda doesn't execute.